File: src/main/java/raylras/zen/util/CSTNodes.java
package raylras.zen.util;

import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.TokenStream;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Queue;

public class CSTNodes {

    public static ParseTree getCstAtPosition(ParseTree root, Position pos) {
        Deque<ParseTree> deque = getCstStackAtPosition(root, pos);
        return deque.peekFirst();
    }

    public static Deque<ParseTree> getCstStackAtPosition(ParseTree root, Position pos) {
        Queue<ParseTree> tempQueue = new ArrayDeque<>();
        tempQueue.add(root);
        Deque<ParseTree> result = new ArrayDeque<>();
        while (!tempQueue.isEmpty()) {
            ParseTree cst = tempQueue.poll();
            Range range = Range.of(cst);
            if (range.contains(pos)) {
                result.addFirst(cst);
                tempQueue.clear();
                for (int i = 0; i < cst.getChildCount(); i++) {
                    tempQueue.add(cst.getChild(i));
                }
            }
        }
        return result;
    }

    public static int getTokenType(Token token) {
        if (token != null) {
            return token.getType();
        } else {
            return Token.INVALID_TYPE;
        }
    }

    public static TerminalNode getPrevTerminal(TokenStream tokenStream, ParseTree node) {
        Token prevToken = getPrevToken(tokenStream, node);
        if (prevToken == null) {
            return null;
        }
        ParseTree root = getRoot(node);
        Range range = Range.of(prevToken);
        ParseTree prevNode = getCstAtPosition(root, range.end());
        return (prevNode instanceof TerminalNode) ? (TerminalNode) prevNode : null;
    }

    public static Token getPrevToken(TokenStream tokenStream, ParseTree node) {
        int i = getStartTokenIndex(node) - 1;
        while (i >= 0) {
            Token token = tokenStream.get(i);
            if (token.getChannel() == Token.DEFAULT_CHANNEL) {
                return token;
            }
            i--;
        }
        return null;
    }

    public static Token getNextToken(TokenStream tokenStream, ParseTree node) {
        int i = getStopTokenIndex(node) + 1;
        while (i <= tokenStream.size()) {
            Token token = tokenStream.get(i);
            if (token.getChannel() == Token.DEFAULT_CHANNEL) {
                return token;
            }
            i++;
        }
        return null;
    }

    public static String getText(ParseTree cst) {
        if (cst != null) {
            return cst.getText();
        } else {
            return "";
        }
    }

    public static <T extends ParseTree> T findParentOfType(ParseTree cst, Class<T> parentClass) {
        ParseTree current = cst;
        while (current != null) {
            if (parentClass.isInstance(current)) {
                return parentClass.cast(current);
            } else {
                current = current.getParent();
            }
        }
        return null;
    }


    public static ParseTree findParentOfTypes(ParseTree cst, Class<?>... parents) {
        ParseTree current = cst;
        while (current != null) {
            for (Class<?> parent : parents) {
                if (parent.isInstance(current)) {
                    return current;
                }
            }
            current = current.getParent();
        }
        return null;
    }

    private static int getStartTokenIndex(ParseTree node) {
        if (node instanceof TerminalNode) {
            return ((TerminalNode) node).getSymbol().getTokenIndex();
        }
        if (node instanceof ParserRuleContext) {
            return ((ParserRuleContext) node).getStart().getTokenIndex();
        }
        return -1;
    }

    private static int getStopTokenIndex(ParseTree node) {
        if (node instanceof TerminalNode) {
            return ((TerminalNode) node).getSymbol().getTokenIndex();
        }
        if (node instanceof ParserRuleContext) {
            return ((ParserRuleContext) node).getStop().getTokenIndex();
        }
        return -1;
    }

    private static ParseTree getRoot(ParseTree node) {
        ParseTree root = node;
        while (root != null && root.getParent() != null) {
            root = root.getParent();
        }
        return root;
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/util/CSTNodes.kt
package raylras.zen.util

import org.antlr.v4.runtime.ParserRuleContext
import org.antlr.v4.runtime.Token
import org.antlr.v4.runtime.TokenStream
import org.antlr.v4.runtime.tree.ParseTree
import org.antlr.v4.runtime.tree.TerminalNode

fun ParseTree?.getTerminalAt(pos: TextPosition): TerminalNode? {
    return getCstStackAt(pos).firstOrNull() as? TerminalNode
}

fun ParseTree?.getCstStackAt(pos: TextPosition): List<ParseTree> {
    val result = ArrayDeque<ParseTree>()
    val tempQueue = mutableListOf(this)
    while (tempQueue.isNotEmpty()) {
        val current = tempQueue.removeFirst() ?: continue
        if (pos in current.textRange) {
            result.addFirst(current)
            tempQueue.clear()
            for (i in 0 until current.childCount) {
                tempQueue.add(current.getChild(i))
            }
        }
    }
    return result
}

fun ParseTree?.walkParent(): Sequence<ParseTree> {
    return generateSequence(this) { it.parent }
}

fun TerminalNode?.getPrev(tokenStream: TokenStream?): TerminalNode? {
    val prevToken = this?.startToken?.getPrev(tokenStream) ?: return null
    val prevNode = this.root.getTerminalAt(prevToken.textRange.end)
    return prevNode
}

fun Token?.getPrev(tokenStream: TokenStream?, channel: Int = Token.DEFAULT_CHANNEL): Token? {
    this ?: return null
    tokenStream ?: return null
    var i = this.tokenIndex.dec()
    while (i >= 0) {
        val token = tokenStream[i]
        if (token.channel == channel) {
            return token
        } else {
            i--
        }
    }
    return null
}

val ParseTree.startToken: Token?
    get() = when (this) {
        is TerminalNode -> {
            this.symbol
        }

        is ParserRuleContext -> {
            this.start
        }

        else -> null
    }

val ParseTree.stopToken: Token?
    get() = when (this) {
        is TerminalNode -> {
            this.symbol
        }

        is ParserRuleContext -> {
            this.stop
        }

        else -> null
    }

val ParseTree.root: ParseTree
    get() {
        var root = this
        while (root.parent != null) {
            root = root.parent
        }
        return root
    }
